docs/proposals/20230612-node-os-image-caching.md

```diff
@@ -0,0 +1,305 @@
+
+#### As an operator of clusters managed by CAPZ, I want to reduce monetary and environmental costs
+
+As a cluster operator I want to have lower monetary and environmental costs. I want my Nodes' OS images to be continually cached, and thus I will be able to avoid longer pull times and avoid having to create warm nodes and overprovision. This will be better for the environment and cheaper for me as I will not have to waste excess resources on warm nodes.
+
+#### As an operator, I want to be able to be more inline with security compliance as I spin up nodes
+
+As an operator I would like to be able to have my node’s OS image cache for the ability to avoid security alerts, cache security updates and patches so that I’ll be more security compliant and up to date when I have to spin up a new node. Otherwise, flags will be raised that a node is out of date on boot up before it finishes patching and a temporary security risk will exist.
+
+### Implementation Details/Notes/Constraints
+
+The plan is to modify the existing Controllers with the Node Prototype Pattern as desired. These controller additions can be added to AzureMachine Controller and AzureMachinePool Controller.
+
+An operator will be able to decide to turn the feature on or off with an environment variable on clusterctl initialization, and then on a cluster by cluster basis can alter the AzureMachineTemplates and AzureMachinePools to specify the use of the feature. They can also update the AzureMachineTemplate and AzureMachinePool to customize how long they want the caching interval to be (see the yaml files below in this section for the caching interval).
+
+Example of the enviornment variable being turned on:
+
+```
+export AZURE_OS_CACHING=true
+```
+
+The controller will maintain a timestamp in each AzureMachinePool and AzureMachineTemplate, and when the current time is the chosen interval ahead or more, the controller will perform the caching. Since the current controller manager requeues all objects every ten minutes by default the objects will be requeued shortly after its due time to be recached. This is because typically we expect to cache every 24 hours and it is very unexpected that this won't be frequent enough considering normal patch rates. When the process is started it should go through the nodes of the cluster, choose a healthy node, shut it down, take a snapshot of it, restart it, create a shared image gallery image, delete the snapshot, and then configure the AzureMachineTemplate specs to use that shared image gallery image. After, it will store the current time as its timestamp.
+
+Example of how the timestamp will be maintained in the AzureMachinePool and AzureMachineTemplates:
+
+```yaml
+status:
+  conditions:
+  - lastTransitionTime: "2023-06-12T23:14:55Z"
```

Comment:
as we discussed offline, `lastTransitionTime` is only updated if there is a change in the condition `status` https://github.com/kubernetes-sigs/cluster-api/blob/main/util/conditions/setter.go#L39. In general, conditions are meant to be user-facing and informational about the operational state of the cluster, not used to communicate timestamps to controllers. You can read more about conditions here: https://github.com/kubernetes-sigs/cluster-api/blob/main/docs/proposals/20200506-conditions.md#constraints

Reply:
Moved to a non-condition in the status to represent this timestamping.